airflow/providers/google/cloud/operators/vertex_ai/custom_job.py

```diff
@@ -180,6 +180,38 @@ def execute(self, context: Context) -> None:
             stacklevel=2,
         )
 
+    def execute_complete(self, context: Context, event: dict[str, Any]) -> dict[str, Any] | None:
+        if event["status"] == "error":
+            raise AirflowException(event["message"])
+        training_pipeline = event["job"]
+        custom_job_id = self.hook.extract_custom_job_id_from_training_pipeline(training_pipeline)
+        self.xcom_push(context, key="custom_job_id", value=custom_job_id)
+        if model_id := self.hook.extract_model_id_from_training_pipeline(training_pipeline):
```

Comment:
Could we do something like this and not update the code for `extract_model_id_from_training_pipeline` function ?
```
        try:
            model = training_pipeline["model_to_upload"]
            model_id = self.hook.extract_model_id(model)
            self.xcom_push(context, key="model_id", value=model_id)
            VertexAIModelLink.persist(context=context, task_instance=self, model_id=model_id)
            return model
        except KeyError:
            self.log.warning(
                "It is impossible to get the Model. "
                "The Training Pipeline did not produce a Managed Model because it was not "
                "configured to upload a Model. Please ensure that the 'model_serving_container_image_uri' "
                "and 'model_display_name' parameters are passed in when creating a Training Pipeline, "
                "and check that your training script saves the model to os.environ['AIP_MODEL_DIR']."
            )
            return None
```

Reply:
What's wrong with updating the function? It is better to move all these exceptions outside of execute for the sake of readability. If you are worried that the hook method could be used somewhere else, then it is unlikely, because the `extract_model_id_from_training_pipeline()` did not exist before, it was added in my last PR that introduced the deferrable mode to the operators.

Reply:
My point was that `extract_` functions are helpers. And these functions should be simple and shouldn't check obj existence and then make a decision what value to return. Also we already have a function which `extract_model_id` from the model object.  Maybe it was wrong to create a new function which does the same, but firstly trying to get an object, which can exist or not. And now we try to make this function more complex to check obj existence before extracting the value.

Reply:
I'm agree with @MaksYermak. If we have some functionality implemented, then we definitely shouldn't duplicate it. Otherwise we bring more complexity without new value and potentially new bugs when somebody from the community modifies only one of these functions.

Reply:
I have updated the PR to address the comment. Please verify @MaksYermak @moiseenkov